asf_heat_pump_suitability/analysis/exploratory/create_full_dataset/create_full_dataset_plymouth.py

```diff
@@ -0,0 +1,1867 @@
+    "s3://asf-heat-pump-suitability/exploration/spatial_clustering_plymouth/Lower_Layer_Super_Output_Area_(2021)_to_LAD_(April_2023)_Lookup_in_England_and_Wales.csv"
+)
+
+# %%
+# Join IMD decile to UPRNs
+imd_df = (
+    imd_df.select(["Reference area", "2019"])
+    .rename({"Reference area": "lsoa_name", "2019": "imd_decile"})
+    .join(
+        lsoa_names_df.select(["LSOA21CD", "LSOA21NM"]),
+        how="left",
+        left_on="lsoa_name",
+        right_on="LSOA21NM",
+    )
+)
+
+features_df = features_df.join(
+    imd_df.select(["LSOA21CD", "imd_decile"]), how="left", on="LSOA21CD"
+)
+
+# %% [markdown]
+# ## Fill missing data - garden size
+#
+# - use average by property type (+ area)
+# - sum up the different garden sizes
+
+# %%
+# Load land registry polygons for Plymouth
+land_extent_gdf = gpd.read_file(
+    "s3://asf-heat-pump-suitability/exploration/spatial_clustering_plymouth/Land_Registry_Cadastral_Parcels.gml",
+    columns=["NATIONALCADASTRALREFERENCE", "geometry"],
+)
+
+# Preprocess land polygons
+land_extent_gdf = geo_utils.transform_gdf_drop_duplicates(land_extent_gdf)
+land_extent_gdf["land_area_m2"] = land_extent_gdf["geometry"].area
+
+# Get Plymouth OS building footprints
+plymouth_building_footprints_gdf = (
+    plymouth_residential_buildings_gdf[["geometry", "ID", "building_area_m2"]]
+    .rename(columns={"ID": "building_id"})
+    .assign(source="os")
+    .copy()
+)
+
+# %%
+# Remove land extent geometries which are not polygons
+print(land_extent_gdf["geometry"].geom_type.value_counts())
+land_extent_gdf = land_extent_gdf[land_extent_gdf.geometry.geom_type == "Polygon"]
+
+# Get intersection of land polygons and building polygons
+intersection_gdf = garden_size.generate_gdf_land_building_overlay(
+    land_parcels_gdf=land_extent_gdf,
+    building_footprints_gdf=plymouth_building_footprints_gdf,
+)
+
+# Get garden size
+gardens_gdf = garden_size.generate_gdf_garden_size(intersection_gdf, land_extent_gdf)
+
+# Join garden size to UPRNs
+uprns_with_garden_gdf = pl.from_pandas(
+    gpd.sjoin(
+        os_uprn_gdf,
+        gardens_gdf,
+        how="inner",
+        predicate="intersects",
+    ).drop(columns=["geometry", "index_right"])
+)
+
+# Identify UPRNs with multiple garden sizes and find the range between min and max garden size calculated for them
+duplicated_gdf = (
+    uprns_with_garden_gdf.filter(pl.col("UPRN").is_duplicated())
+    .group_by("UPRN")
+    .agg(pl.min("garden_area_m2").alias("min"), pl.max("garden_area_m2").alias("max"))
+    .with_columns((pl.col("max") - pl.col("min")).alias("range"))
+)
+
+# Deduplicate UPRNs to the smallest garden size - the join allows us to retain the other columns in the df
+# TODO - use pl.over instead as it's more efficient
+uprns_with_garden_gdf = uprns_with_garden_gdf.join(
+    uprns_with_garden_gdf.group_by("UPRN").agg(pl.min("garden_area_m2")),
+    how="left",
+    on="UPRN",
+).filter(pl.col("garden_area_m2") == pl.col("garden_area_m2_right"))
```

Comment:
is something like this possible? feels more intuitive when we read it

```suggestion
uprns_with_garden_gdf = uprns_with_garden_gdf.join(
    uprns_with_garden_gdf.group_by("UPRN").agg(pl.min("garden_area_m2").alias("min_garden_area_m2")),
    how="left",
    on="UPRN",
).filter(pl.col("garden_area_m2") == pl.col("garden_area_m2_right"))
```

Reply:
but also, if for some reason there are multiple lines for the same UPRN with the same garden size, won't we have the same UPRN multiple times? so, basically not truly deduplicating?

Reply:
yes i think i did check and this doesnt happen (i would need to check again to be sure) but great point, this needs to be fixed when refactoring.